### PR TITLE
cmd/systray: add exit nodes menu

### PR DESCRIPTION
**Pull request by willnorris**
This commit builds the exit node menu including the recommended exit node, if available, as well as tailnet and mullvad exit nodes.

This does not yet update the menu based on changes in exit node outside of the systray app, which will come later.  This also does not include the ability to run as an exit node.

Updates #1708

Change-Id: Ia101a4a3005adb9118051b3416f5a64a4a45987d

### CODE REVIEW

**Comment by willnorris** [comment]
Actually the country flags are trivial to add as well, so doing that now.

**Comment by willnorris** [comment]
this also doesn't yet update the systray app icon to include the arrow indicator that an exit node is in use.  I'm gonna have to take some time to draw a proper arrow with the graphics library we're using.  (https://github.com/fogleman/gg/issues/138)

**Comment by willnorris** [comment]
GNOME won't expand the country menus. Given that they all open downward, instead of as flyout menus, I'm guessing there may be a max-depth.  If so, I'll probably update this to treat all countries as single-city on GNOME.

Screenshot From 2024-12-20 16-13-07

File: cmd/systray/systray.go [hunk 1]
```diff
@@ -375,6 +402,466 @@ func sendNotification(title, content string) {
...
+	"aq": "🇦🇶",
+	"ar": "🇦🇷",
+	"as": "🇦🇸",
+	"at": "🇦🇹",
+	"au": "🇦🇺",
+	"aw": "🇦🇼",
+	"ax": "🇦🇽",
+	"az": "🇦🇿",
+	"ba": "🇧🇦",
+	"bb": "🇧🇧",
```
Comment (bradfitz) [inline cmd/systray/systray.go:629]: This is all unnecessary.

You can build the emoji programmatically from the two letters: https://go.dev/play/p/rKXgwaHhRAG

Comment (willnorris) [reply cmd/systray/systray.go:629]: oooh!  TIL

File: cmd/systray/systray.go [hunk 2]
```diff
@@ -375,6 +402,230 @@ func sendNotification(title, content string) {
...
+	noExitNodeMenu := menu.exitNodes.AddSubMenuItemCheckbox("None", "", status.ExitNodeStatus == nil)
+	onClick(noExitNodeMenu, "")
+
+	// Show recommended exit node if available.
+	if status.Self.CapMap.Contains(tailcfg.NodeAttrSuggestExitNodeUI) {
+		sugg, err := localClient.SuggestExitNode(ctx)
+		if err == nil {
+			title := "Recommended: "
+			if loc := sugg.Location; loc.Valid() && loc.Country() != "" {
+				flag := countryFlag(strings.ToUpper(loc.CountryCode()))
```
Comment (bradfitz) [inline cmd/systray/systray.go:437]: no need for the strings.ToUpper; I accounted for that already in the countryFlag func

(here and elsewhere)

File: cmd/systray/systray.go [hunk 3]
```diff
@@ -375,6 +402,230 @@ func sendNotification(title, content string) {
...
+		}()
+	}
+
+	noExitNodeMenu := menu.exitNodes.AddSubMenuItemCheckbox("None", "", status.ExitNodeStatus == nil)
+	onClick(noExitNodeMenu, "")
+
+	// Show recommended exit node if available.
+	if status.Self.CapMap.Contains(tailcfg.NodeAttrSuggestExitNodeUI) {
+		sugg, err := localClient.SuggestExitNode(ctx)
+		if err == nil {
```
Comment (bradfitz) [inline cmd/systray/systray.go:434]:
```go
if err != nil {
   // Ignore.
   continue
}
// outdented code
```

Comment (bradfitz) [reply cmd/systray/systray.go:434]: oh, it's not in a loop, nevermind.

File: cmd/systray/systray.go [hunk 4]
```diff
@@ -375,6 +402,230 @@ func sendNotification(title, content string) {
...
+			}
+		}
+	}
+
+	// TODO: "Allow Local Network Access" and "Run Exit Node" menu items
+}
+
+// mullvadPeers contains all mullvad peer nodes, sorted by country and city.
+type mullvadPeers struct {
+	countries map[string]*mvCountry // country code => country
```
Comment (bradfitz) [inline cmd/systray/systray.go:537]: country code uppercase => ... ?

might be worth a type where you can clarify it's ISO 3166-1 alpha-2 if it is.

but it's also not exported, so meh

File: cmd/systray/systray.go [hunk 5]
```diff
@@ -375,6 +402,230 @@ func sendNotification(title, content string) {
...
+}
+
+// sortedCountries returns countries containing mullvad nodes, sorted by name.
+func (mp mullvadPeers) sortedCountries() []*mvCountry {
+	countries := make([]*mvCountry, 0, len(mp.countries))
+	for _, country := range mp.countries {
+		countries = append(countries, country)
+	}
+	sort.Slice(countries, func(i, j int) bool {
```
Comment (bradfitz) [inline cmd/systray/systray.go:546]: the cool way nowadays is [slices.Sort](https://pkg.go.dev/slices#SortFunc) and then `return cmp.Compare` (https://pkg.go.dev/cmp#Compare)

File: cmd/systray/systray.go (hunk 3 above)
Comment (willnorris) [reply cmd/systray/systray.go:434]: this isn't in a loop, so can't continue

File: cmd/systray/systray.go [hunk 6]
```diff
@@ -375,6 +402,230 @@ func sendNotification(title, content string) {
...
+type mullvadPeers struct {
+	countries map[string]*mvCountry // country code => country
+}
+
+// sortedCountries returns countries containing mullvad nodes, sorted by name.
+func (mp mullvadPeers) sortedCountries() []*mvCountry {
+	countries := make([]*mvCountry, 0, len(mp.countries))
+	for _, country := range mp.countries {
+		countries = append(countries, country)
+	}
```
Comment (bradfitz) [inline cmd/systray/systray.go:545]: `countries := slices.Clone(mp.countries)`

File: cmd/systray/systray.go [hunk 7]
```diff
@@ -375,6 +402,230 @@ func sendNotification(title, content string) {
...
+// sortedCities returns cities containing mullvad nodes, sorted by name.
+func (mc *mvCountry) sortedCities() []*mvCity {
+	cities := make([]*mvCity, 0, len(mc.cities))
+	for _, city := range mc.cities {
+		cities = append(cities, city)
+	}
+	sort.Slice(cities, func(i, j int) bool {
+		return cities[i].name < cities[j].name
+	})
+	return cities
```
Comment (bradfitz) [inline cmd/systray/systray.go:568]: same

**Review by bradfitz — APPROVED** [review]

File: cmd/systray/systray.go (hunk 3 above)
Comment (willnorris) [reply cmd/systray/systray.go:434]: I've got some pretty deeply nested loops later on though that could probably be simplified by continuing early.

File: cmd/systray/systray.go (hunk 2 above)
Comment (willnorris) [reply cmd/systray/systray.go]: done

File: cmd/systray/systray.go (hunk 4 above)
Comment (willnorris) [reply cmd/systray/systray.go]: added comment.  leaving off the type for now

File: cmd/systray/systray.go (hunk 5 above)
Comment (willnorris) [reply cmd/systray/systray.go]: done

File: cmd/systray/systray.go (hunk 6 above)
Comment (willnorris) [reply cmd/systray/systray.go]: `mp.countries` is a map, not a slice so can't simply clone.  I did switch this to `countries := slices.Collect(maps.Values(mp.countries))` though.

**Pull request merged**